dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/scheduling/ContinuousAnalyticsTableJob.java

```diff
@@ -95,9 +95,17 @@ public void execute( JobConfiguration jobConfiguration )
             DEFAULT_HOUR_OF_DAY );
 
         Date now = new Date();
+
         Date defaultNextFullUpdate = DateUtils.getNextDate( fullUpdateHourOfDay, now );
-        Date nextFullUpdate = (Date) systemSettingManager.getSystemSetting( SettingKey.NEXT_ANALYTICS_TABLE_UPDATE,
-            defaultNextFullUpdate );
+
+        Date nextFullUpdate = (Date) systemSettingManager.getSystemSetting( SettingKey.NEXT_ANALYTICS_TABLE_UPDATE );
```

Comment:
These two are semantically slightly different. The default value passed to `getSystemSetting` will only be used if there is no database entry for the setting. If there is a setting but its value is `null` you would get a `null` returned. 

As Lars said what we should answer is if this is a problem that has been introduced causing this or if we e.g. had bad data in the database (entry with `null` value) caused this.

Reply:
Just wondering if that's the case, seems we check if the `value` is null or if I read this wrongly:

https://github.com/dhis2/dhis2-core/blob/master/dhis-2/dhis-services/dhis-service-setting/src/main/java/org/hisp/dhis/setting/DefaultSystemSettingManager.java#L203

https://github.com/dhis2/dhis2-core/blob/master/dhis-2/dhis-services/dhis-service-setting/src/main/java/org/hisp/dhis/setting/SystemSetting.java#L68

Reply:
@larshelge, @jbee : I agree with you guys, it is just quick fix. I informed Phil I have to go deeper which takes a bit more time

Reply:
Had a look at the loading code itself. It is interesting that we have `hasValue()` which is checking `value != null` but then we use `displayValue` - Can have an entry with `value` being non `null` while `displayValue` is `null`? This seems to open some possibilities for error or unexpected behaviour.

Reply:
Ugh the `displayValue` is news to me, not sure when that was added. I guess that is for potential translations of system setting values. Not sure if that was a great idea.

Reply:
This part seems like it is prone to race conditions too..?

https://github.com/dhis2/dhis2-core/blob/master/dhis-2/dhis-services/dhis-service-setting/src/main/java/org/hisp/dhis/setting/SystemSetting.java#L130

Reply:
I found only one possibility how this can happen:
1. In the database there is systemsettings table (name -> regular value, translation -> translated value)
2.  Date nextFullUpdate = (Date) systemSettingManager.getSystemSetting( SettingKey.NEXT_ANALYTICS_TABLE_UPDATE,
            defaultNextFullUpdate );
getSystemSetting is looking in the cache and if the value is not present defualtNextFullUpdate is applied:
SerializableOptional value = settingCache.get( key.getName(),
            k -> getSystemSettingOptional( k, defaultValue ) ).get();
3. getSystemSettingOptional looks like this:`if ( setting != null && setting.hasValue() )
        {
            if ( isConfidential( name ) )
            {
                try
                {
                    return SerializableOptional.of( pbeStringEncryptor.decrypt( (String) setting.getDisplayValue() ) );
                }
                catch ( EncryptionOperationNotPossibleException e )
                {
                    log.warn( "Could not decrypt system setting '" + name + "'" );
                    return SerializableOptional.empty();
                }
            }
            else
            {

                return SerializableOptional.of( setting.getDisplayValue() );
            }
        }`

Reply:
@larshelge FYI: Had a call with @d-bernat and we concluded that this `null` is a result of database having a `value` of string `"null"` which is understood as JSON and converted to java yielding `null` but before that conversion happens the checks for existence of a value will determine that there is a value present. That is why we do not use the default value provided. This seems wrong. However, changing it could affect other settings so I think it is a good idea to do this work-around for the patch and backports just for this job and do another PR where we change the general behaviour for `master` (2.38)

Reply:
Ah interesting. Good investigation.

I am still reluctant to fix this in _just one place_. We will get the "null" string problem many other places then. Better to fix this centrally and test well.

We could even release a SQL query to correct the database, though it is hard to work with serializable/binary properties in SQL.

Reply:
@larshelge If you are happy to take the risk to backport the general fix that might affect more places we can just do that :)

Reply:
Right, I think I still would prefer to fix the "root cause" and not the "symptoms". I think though it's been like this for a long time (right?) so it wouldn't be a blocker for 2.37? Instead we could fix this properly at the root in master, let it mature and test properly, and then backport.

Reply:
>  I think though it's been like this for a long time

I'd think so.
What we have not identified is why `"null"` string ends up in the database and if that is a newer issue. But my assumption is that this is an artefact of json conversion that also has been there quite a while. Leaves the question why the continuous analytics job was now affected by it. Could it be as simple as someone editing the job configuration for what then became the demo database snapshot?